File: solutions/Max Area of Island/Max Area of Island.ts
function maxAreaOfIsland(grid: number[][]): number {
    const [M,N]: [number, number] = [grid.length, grid[0].length]
    let maxArea: number = 0
    
    for (let i=0; i<M; i++) {
        for (let j=0; j<N; j++) {
            if (grid[i][j] === 1) {
                maxArea = Math.max(maxArea, dfs(i,j,0))

            }
        }
    }

    return maxArea

    ///////////////
    // func defs //
    ///////////////
    function dfs(x: number, y: number, curArea: number): number {
        // base cases
        if (
            x < 0
            || y < 0 
            || x >= M
            || y >= N
            || grid[x][y] === 0
        ) return 0;

        // ops
        grid[x][y] = 0
        let thisArea: number = 1;

        // recursion
        thisArea += dfs(x+1,y,curArea+1)
        thisArea += dfs(x-1,y,curArea+1)
        thisArea += dfs(x,y+1,curArea+1)
        thisArea += dfs(x,y-1,curArea+1)

        return thisArea;
    }
};

File: solutions/Max Area of Island/Max Area of Island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        
        maxArea = 0
        curArea = 0

        dq = deque()
        def BFS(x: int, y: int) -> None:
            nonlocal curArea

            dq.append([x,y])
            grid[x][y] = 0

            while dq:
                cx, cy = dq.popleft()
                curArea += 1

                for nx,ny in [
                    [cx+1,cy],
                    [cx-1,cy],
                    [cx,cy+1],
                    [cx,cy-1]
                ]:
                    if 0 <= nx < m and 0 <= ny < n and grid[nx][ny] == 1:
                        dq.append([nx,ny])
                        grid[nx][ny] = 0

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1: # LAND HO!
                    BFS(i,j)
                    maxArea = max(maxArea, curArea)
                    curArea = 0

        return maxArea